src/validator/types/datetime.rs

```diff
@@ -0,0 +1,89 @@
+use chrono;
+use lazy_static::lazy_static;
+use regex::Regex;
+use serde_json::Value;
+
+use crate::validator::{scope::ScopedSchema, state::ValidationState, types::validate_as_string};
+
+lazy_static! {
+    // ajv v6.7.0 compatible
+    // https://github.com/epoberezkin/ajv/blob/v6.7.0/lib/compile/formats.js#L90
+    static ref DATE_REGEX: Regex =
+        Regex::new(r"^(\d\d\d\d)-(\d\d)-(\d\d)$").unwrap();
+
+    // ajv v6.7.0 compatible
+    // https://github.com/epoberezkin/ajv/blob/v6.7.0/lib/compile/formats.js#L104
+    static ref TIME_REGEX: Regex =
+        Regex::new(r"^(\d\d):(\d\d):(\d\d)(\.\d+)?(z|[+-]\d\d:\d\d)?$").unwrap();
+}
+
+pub fn validate_as_datetime(scope: &ScopedSchema, data: &Value) -> ValidationState {
+    let mut state = validate_as_string(scope, data);
+
+    if state.is_valid()
+        && chrono::DateTime::parse_from_rfc3339(data.as_str().expect("invalid validate_as_string")).is_err()
+    {
+        state.push_error(scope.error("type", "unable to parse as 'datetime'"));
+    }
+
+    state
+}
+
+fn is_leap_year(year: usize) -> bool {
+    // https://tools.ietf.org/html/rfc3339#appendix-C
+    year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)
+}
+
+fn days(year: usize, month: usize) -> usize {
+    const DAYS: [usize; 12] = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
+
+    if month == 2 && is_leap_year(year) {
+        29
+    } else {
+        DAYS[month - 1]
+    }
+}
+
+pub fn validate_as_date(scope: &ScopedSchema, data: &Value) -> ValidationState {
```

Comment:
would it make sense to keep `pub fn`s at the top ?

Reply:
Don't have a strong opinion here. I prefer module flow in a way that if I have `X`, `Y` and `X` depends / uses `Y`, then the `Y` is first and then `X` (doesn't matter if it's a `pub` or not).

Reply:
I tend to have the `what does this module do` at the top, my thinking being that if someone does not know what is this all about they can read the first few functions and have a high-level overview and if they need to know the details then they can keep reading, and the farther you are into the file the more detailed/low-level the code becomes.

Reply:
The issue you have is not about fn ordering, but lack of the documentation :)

Reply:
I think that good documentation would not hurt anyone here :P
I also think that having this consistent throughout the project is more important than which style that is - I think it is consistent now ?

May be worth mentioning the in developer documentation somewhere what is the general rule of ordering things inside of a file - this could help people read the code.

Reply:
> I think it is consistent now

Yes, it is.

> May be worth mentioning in the developer documentation somewhere what is the general rule of ordering things inside of a file - this could help people read the code.

I tend to disagree here. All these files have like 150 LOC (validator) and most of them has like < 50 LOC. And if you can't get a clue from 50 LOC with features like Structured / Outline View (IDE features like CLion, VS Code, ...), etc. than I'm sorry. Documentation won't help here.